**ORE SPESE SU QUESTO HOMEWORK**\
- 2 ora e mezza il 14-12
- 1 ora il 15-12
- 1 ora e mezza il 16-12

In [332]:
import pandas as pd
import numpy as np
import networkx as nx

### Let's import the datasets
**HERO-NETWORK:**\
It contains the network of heroes who have appeared together in the comics.

In [317]:
dfheros = pd.read_csv(r'C:\Users\gianl\Documents\Università_4°anno(2022-2023\ADM\HOMEWORK\HOMEWORK 5\hero-network.csv')

In [299]:
dfheros.head(5)

hero1                 hero2
0         LITTLE, ABNER        PRINCESS ZANDA
1         LITTLE, ABNER  BLACK PANTHER/T'CHAL
2  BLACK PANTHER/T'CHAL        PRINCESS ZANDA
3         LITTLE, ABNER        PRINCESS ZANDA
4         LITTLE, ABNER  BLACK PANTHER/T'CHAL

In [300]:
dfheros.shape

(574467, 2)

**EDGES:**\
It contains two columns (hero, comic), indicating which comics the heroes appear in.

In [301]:
dfedges = pd.read_csv(r'C:\Users\gianl\Documents\Università_4°anno(2022-2023\ADM\HOMEWORK\HOMEWORK 5\edges.csv')

In [302]:
dfedges.head(5)

hero   comic
0  24-HOUR MAN/EMMANUEL  AA2 35
1  3-D MAN/CHARLES CHAN   AVF 4
2  3-D MAN/CHARLES CHAN   AVF 5
3  3-D MAN/CHARLES CHAN   COC 1
4  3-D MAN/CHARLES CHAN  H2 251

In [303]:
dfedges.shape

(96104, 2)

**NODES:**\
It contains two columns (node, type), indicating the nodes' name and type (comic, hero).

In [304]:
dfnodes = pd.read_csv(r'C:\Users\gianl\Documents\Università_4°anno(2022-2023\ADM\HOMEWORK\HOMEWORK 5\nodes.csv')

In [305]:
dfnodes.head(5)

node   type
0               2001 10  comic
1                2001 8  comic
2                2001 9  comic
3  24-HOUR MAN/EMMANUEL   hero
4  3-D MAN/CHARLES CHAN   hero

In [306]:
dfnodes.shape

(19090, 2)

# QUESTION 1: DATA

## GRAPHS SETUP
### 1)

An edge between two heroes can be found if they have appeared in the same comic together. The number of edges between two heroes represents the number of times they have collaborated in different comics. Let's find all edges

In [333]:
G_heros = nx.MultiGraph()

In [335]:
primo = heros.apply(lambda row:G_heros.add_edge(row['hero1'], row['hero2']),axis=1)

### 2)

## DATA PREPROCESSING
### 1)

**1.1**

"Some heroes' names in 'hero-netowrk.csv' have extra spaces at the end of their names compared to their names in 'edges.csv'" :

The following command gets the string of heroes inside column 'hero1' (in dataset heros) that have a space in the last position: '  '. Once we have these heroes, we edit the string by deleting the space in the last position.

In [318]:
for i in range(len(dfheros.hero1)):
    if dfheros.hero1[i][-1] == ' ':
        dfheros.hero1[i]= dfheros.hero1[i][0:-1]

**1.2**

Some heroes' names in 'hero-netowrk.csv' have an extra '/' at the end of their names compared to their names in 'edges.csv'.

- We could do the same things of above, but we noticed that in the column hero (dataset edges) we have heros which have in the last position a slash: '/'. So we can't remove these heros but we have to remove only the heroes that within hero1 (in dataset heros) have a slash in the last position, while in hero (dataset edges) they don't.

- However, to make things easier, we decided to remove the / in the last position in all heroes even those that are normally registered with the / in the last position (for example 8-BALL/ it is normally registered with the / in the last position because it has it in both the heros and edges datasets, but to make things easier we will remove the / even for 8-BALL/ in both dataset)

- The following query, therefore, removes the / in last position (to whoever has it) to all heroes in 'hero1' (dataset heros) and in 'hero' (dataset edges):

In [308]:
#for the dataset 'dfheros'
for i in range(len(dfheros.hero1)):
    if dfheros.hero1[i][-1] == '/':
        dfheros.hero1[i]= dfheros.hero1[i][0:-1]
        
#for the dataset 'dfedges'
for i in range(len(dfedges.hero)):
    if dfedges.hero[i][-1] == '/':
        dfedges.hero[i]= dfedges.hero[i][0:-1]

**1.3**

The hero name 'SPIDER-MAN/PETER PARKER' in 'edges.csv' has been changed to 'SPIDER-MAN/PETER PAR' in 'hero-network.csv' due to a string length limit in 'hero-network.csv'

We can see inside the dataset edges there are 1577 times the name 'SPIDER-MAN/PETER PARKER' in the column 'hero'.

In [183]:
dfedges[dfedges.hero == 'SPIDER-MAN/PETER PARKER'].shape

(1577, 2)

In order to avoid the problem of having two different name for the same hero in two different dataset, we decided to change inside the dataset edges, the name of the hero 'SPIDER-MAN/PETER PARKER' in 'SPIDER-MAN/PETER PAR'.

In [185]:
dfedges.hero = dfedges.hero.replace('SPIDER-MAN/PETER PARKER', 'SPIDER-MAN/PETER PAR')

Altought before were 1577 rows, now we can see that 'SPIDER-MAN/PETER PARKER' doesn't appear anymore.

In [186]:
dfedges[dfedges.hero == 'SPIDER-MAN/PETER PARKER'].shape

(0, 2)

### 2)

Some entries in the 'hero-network.csv' have the same hero in both columns. In the graph, these entries form a self-loop. Because a self-loop makes no sense in this network, you can safely remove those from the dataset.

We can observe from the following cell that the rows we need to delete are 2232.

In [269]:
counter = 0
for i in range(len(dfheros.hero1)):
    if dfheros.hero1[i] == dfheros.hero2[i]:
        counter += 1
print('The number of total rows are: ', len(dfheros), "\nInstead, the number of rows that have the same hero in 'hero1' and 'hero2' are: ", counter)

The number of total rows are:  574467 
Instead the number of lines that have the same hero in 'hero1' and 'hero2' are:  2232


The following code iterates through the heros dataframe, if it encounters a row that has the same name in 'hero1' and 'hero2' then it deletes that row.

In [270]:
for i in range(len(dfheros.hero1)):
    if dfheros.hero1[i] == dfheros.hero2[i]:
        dfheros = dfheros.drop(i)

Let's verify that 2232 rows have actually been deleted!

Before we had a length of 574467, subtracting 2232 from it, we should obtain that the length of the dataframe is now equal to 572235:

In [271]:
len(dfheros)

572235

# QUESTION 2: BACKEND IMPLEMENTATION

## Functionality 1 - extract the graph's features

#### F1.1 The number of nodes in the network (if type 2, report for both node types)

#### F1.2 The number of collaborations of each superhero with the others (only if type 1

In [336]:
primo

0          8
1          7
2          6
3          9
4          8
          ..
574462    55
574463    45
574464    17
574465    23
574466     3
Length: 574467, dtype: int64

In [342]:
dfheros

hero1                 hero2
0              LITTLE, ABNER        PRINCESS ZANDA
1              LITTLE, ABNER  BLACK PANTHER/T'CHAL
2       BLACK PANTHER/T'CHAL        PRINCESS ZANDA
3              LITTLE, ABNER        PRINCESS ZANDA
4              LITTLE, ABNER  BLACK PANTHER/T'CHAL
...                      ...                   ...
574462  COLOSSUS II/PETER RA              CALLISTO
574463              CALLISTO               ROGUE /
574464              CALLISTO              CALIBAN/
574465              CALIBAN/               ROGUE /
574466  HULK/DR. ROBERT BRUC     MARKS, DR. SHIELA

[574467 rows x 2 columns]